frontend/dockerfile/docs/reference.md

```diff
@@ -1160,41 +1160,96 @@ The available `[OPTIONS]` are:
 
-To add all files in the root of the build context starting with "hom":
+You can specify multiple source files or directories with `ADD`. The last
+argument must always be the destination. For example, to add two files,
+`file1.txt` and `file2.txt`, from the build context to `/usr/src/things/` in
+the build container:
 
 ```dockerfile
-ADD hom* /mydir/
+ADD file1.txt file2.txt /usr/src/things/
 ```
 
-In the following example, `?` is a single-character wildcard, matching e.g. "home.txt".
+If you specify multiple source files, either directly or using a wildcard, then
+the destination must be a directory (must end with a slash `/`).
+
+To add files from a remote location, you can specify a URL or the address of a
+Git repository as the source. For example:
 
 ```dockerfile
-ADD hom?.txt /mydir/
+ADD https://example.com/archive.zip /usr/src/things/
+ADD git@github.com:user/repo.git /usr/src/things/
 ```
 
-The `<dest>` is an absolute path, or a path relative to `WORKDIR`, into which
-the source will be copied inside the destination container.
+BuildKit detects the type of `<src>` and processes it accordingly.
+
+- If `<src>` is a local file or directory, the contents of the directory are
+  copied to the specified destination. See [Adding files from the build context](#adding-files-from-the-build-context).
+- If `<src>` is a local tar archive, it is decompressed and extracted to the
+  specified destination. See [Adding local tar archives](#adding-local-tar-archives).
+- If `<src>` is a URL, the contents of the URL are downloaded and placed at
+  the specified destination. See [Adding files from a URL](#adding-files-from-a-url).
+- If `<src>` is a Git repository, the repository is cloned to the specified
+  destination. See [Adding files from a Git repository](#adding-files-from-a-git-repository).
+
+#### Adding files from the build context
+
+Any relative or local path that doesn't begin with a `http://`, `https://`, or
+`git@` protocol prefix is considered a local file path. The local file path is
+relative to the build context. For example, if the build context is the current
+directory, `ADD file.txt /` adds the file at `./file.txt` to the root of the
+filesystem in the build container.
+
+When adding source files from the build context, their paths are interpreted as
+relative to the root of the context. If you specify a relative path leading
+outside of the build context, such as `ADD ../something /something`, parent
+directory paths are stripped out automatically. The effective source path in
+this example becomes `ADD something /something`.
 
-The example below uses a relative path, and adds "test.txt" to `<WORKDIR>/relativeDir/`:
+If the source is a directory, the contents of the directory are copied,
+including filesystem metadata. The directory itself isn't copied, only its
+contents. If it contains subdirectories, these are also copied, and merged with
+any existing directories at the destination. Any conflicts are resolved in
```

Comment:
> Any conflicts are resolved in

Not worth complicating probably, but this is true if files are of the same type. If you have a directory and replace it with a regular file, you get an error.

Reply:
I can add this, just to avoid any confusion.